plugins/elantp/fu-elantp-i2c-device.c

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (C) 2020 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include <linux/i2c-dev.h>
+
+#include "config.h"
+
+#include "fu-elantp-common.h"
+#include "fu-elantp-i2c-device.h"
+#include "fu-chunk.h"
+
+struct _FuElantpI2cDevice {
+	FuUdevDevice		 parent_instance;
+	guint16			 iap_ctrl;
+	guint16			 ic_page_count;
+};
+
+G_DEFINE_TYPE (FuElantpI2cDevice, fu_elantp_ic2_device, FU_TYPE_UDEV_DEVICE)
+
+static gboolean
+fu_elantp_i2c_device_send_cmd (FuElantpI2cDevice *self,
+			       guint8 *tx, gssize txsz,
+			       guint8 *rx, gssize rxsz,
+			       GError **error)
+{
+	if (!fu_udev_device_pwrite_full (FU_UDEV_DEVICE (self), 0, tx, txsz, error))
+		return FALSE;
+	if (rxsz == 0)
+		return TRUE;
+	if (!fu_udev_device_pread_full (FU_UDEV_DEVICE (self), 0, rx, rxsz, error))
+		return FALSE;
+	return TRUE;
+}
+
+G_GNUC_UNUSED static gboolean
+fu_elantp_i2c_device_write_cmd (FuElantpI2cDevice *self, guint16 reg, guint16 cmd, GError **error)
+{
+	guint8 buf[4];
+	fu_common_write_uint16 (buf + 0x0, reg, G_LITTLE_ENDIAN);
+	fu_common_write_uint16 (buf + 0x2, cmd, G_LITTLE_ENDIAN);
+	if (g_getenv ("FWUPD_ELANTP_VERBOSE") != NULL)
+		fu_common_dump_raw (G_LOG_DOMAIN, "WriteCmd", buf, sizeof(buf));
+	return fu_elantp_i2c_device_send_cmd (self, buf, sizeof(buf), NULL, 0, error);
+}
+
+G_GNUC_UNUSED static gboolean
+fu_elantp_i2c_device_read_cmd (FuElantpI2cDevice *self, guint16 reg,
+			       guint8 *rx, gsize rxsz, GError **error)
+{
+	guint8 buf[2];
+	fu_common_write_uint16 (buf + 0x0, reg, G_LITTLE_ENDIAN);
+	if (g_getenv ("FWUPD_ELANTP_VERBOSE") != NULL)
+		fu_common_dump_raw (G_LOG_DOMAIN, "ReadCmd", buf, sizeof(buf));
+	return fu_elantp_i2c_device_send_cmd (self, buf, sizeof(buf),
+					      rx, rxsz, error);
+}
+
+static gboolean
+fu_elantp_i2c_device_ensure_iap_ctrl (FuElantpI2cDevice *self, GError **error)
+{
+	return TRUE;
+}
+
+G_GNUC_UNUSED static gboolean
+fu_elantp_i2c_device_write_fw_block (FuElantpI2cDevice *self,
+				     const guint8 *raw_data,
+				     guint16 checksum,
+				     GError **error)
+{
+	guint8 buf[FW_PAGE_SIZE + 4];
+
+	buf[0] = ETP_I2C_IAP_REG_L;
+	buf[1] = ETP_I2C_IAP_REG_H;
+	memcpy (buf + 2, raw_data, FW_PAGE_SIZE);
+	fu_common_write_uint16 (buf + FW_PAGE_SIZE + 2, checksum, G_LITTLE_ENDIAN);
+
+	if (!fu_elantp_i2c_device_send_cmd (self, buf, sizeof(buf), 0, 0, error))
+		return FALSE;
+	g_usleep (35 * 1000);
+	if (!fu_elantp_i2c_device_ensure_iap_ctrl (self, error))
+		return FALSE;
+	if (self->iap_ctrl & (ETP_FW_IAP_PAGE_ERR | ETP_FW_IAP_INTF_ERR)) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_WRITE,
+			     "IAP reports failed write: 0x%x",
+			     self->iap_ctrl);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_elantp_ic2_device_open (FuUdevDevice *device, GError **error)
+{
+//	FuElantpI2cDevice *self = FU_ELANTP_I2C_DEVICE (device);
+	gint addr = 0x15;
```

Comment:
wait, does this mean that all elan touchpads are always hardcoded to I2C address `0x15`?

Reply:
I believe so. Is that the sort of thing that belongs in a quirk?

Reply:
At least something that _supports_ a quirk, even if the default is `0x15` it would allow for designs to use Elan on a different I2C address.

Reply:
Yes, elan touchpad has some other i2c addresses.

Reply:
Punted for now, lets get the HID stuff upstream first.